frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -252,6 +259,54 @@ impl<T: Config> Pallet<T> {
 		}
 	}
 
+	/// Greedily reduce the size of the solution to fit into the block w.r.t length.
+	///
+	/// The length of the solution is largely a function of the number of voters. The number of
+	/// winners cannot be changed, and the `ElectionSize` is a proxy for the total number of stakers.
+	///
+	/// Thus, to reduce the solution size, we need to strip voters.
+	///
+	/// Note that this solution is already computed, and winners are elected based on the merit of
+	/// the total stake in the system. Nevertheless, some of the voters may be removed here.
+	///
+	/// The score must be computed **after** this step. If this step reduces the score too much,
+	/// then the solution must be discarded.
+	pub fn trim_compact_for_length(
+		max_allowed_length: u32,
+		mut compact: CompactOf<T>,
+		voter_index: impl Fn(&T::AccountId) -> Option<CompactVoterIndexOf<T>>,
+	) -> Result<CompactOf<T>, MinerError> {
+		// grab all voters and sort them by least stake.
+		let RoundSnapshot { voters, .. } =
+			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
+		let mut voters_sorted = voters
+			.into_iter()
+			.map(|(who, stake, _)| (who.clone(), stake))
+			.collect::<Vec<_>>();
+		voters_sorted.sort_by_key(|(_, y)| *y);
+		voters_sorted.reverse();
+
+		// we run through the reduction process twice: once using the size_hint function, and once
```

Comment:
Again, given that this is OCW, I won't annoy you now, but we can do this much better. We can _statically_ know the encoded size of each voter, based on how many votes it is casting (e.g. is it a votes1 or a votes5 field in `compact`), so there's not much need for any of this re-encoding at the end of the day.

Reply:
I assume that that precise calculation is embedded in the generated `size_hint` implementation. Actually encoding it is just a hedge against the case where some type has a broken `size_hint` impl.

Reply:
and where is the size_hint currently coming from? 

PS. this business of array of fns seems quite dope. What is the semantics?

Reply:
`Config::CompactSolution` has trait bound `Codec`, which has trait bound `Encode`, which provides [`size_hint`](https://substrate.dev/rustdocs/v3.0.0/parity_scale_codec/trait.Encode.html#method.size_hint). It seems fair to assume that the `size_hint` is accurate most of the time when it's derived, but we can't rule out the possibility of a malformed manual implementation.

Reply:
I def. want to see tests for this, sounds cryptic to me and (somewhat similar to our discussion about `set` vs `mutate` :P) based on assumptions that I am not sure about. 

>  It seems fair to assume that the size_hint is accurate most of the time when it's derived

Reply:
The test would be: 

```
// build a compact as such: 
let test = TestCompact<Voter, Target, PerThing> {
  votes1: vec![(v1, t1), (v2, t2)] // `Voter + Target` per eleement
  votes2: vec![(v1, (t2, p1), t1)] //  `Voter + 2 * Target + PerThing` per element
  votes3: vec![] // `Voter + 3 * Target + 2 * PerThing` per element
  // no need to test votes3 onwards as it has the same logic, but feel free to explore

assert_eq!(test.len(), <correct_value>)
```

Probably should live in `sp-npos-elections/src/tests`

If the `size_hint` cannot uphold to our expectations, please implement a `fn encoded_size(&self) -> usize` or sth similar for this. 

I think

Reply:
I think that d41b07f42 takes care of this for now.